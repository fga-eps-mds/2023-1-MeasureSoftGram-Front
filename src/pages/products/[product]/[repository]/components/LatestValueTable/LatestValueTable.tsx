import React from 'react';
import { format } from 'date-fns';
import { ptBR } from 'date-fns/locale';

import {
  Box,
  TableContainer,
  Table,
  TableCell,
  TableHead,
  TableRow,
  TableBody,
  Paper,
} from '@mui/material';

import { Historical } from '@customTypes/repository';

  
function createData(
  name: string,
  latestValue: number,
  latestData: string,
) {
  return { name, latestValue, latestData};
}

interface Prop {
    title: string;
    latestValue: Historical[];
}

function LatestValueTable({title, latestValue}: Prop) {    
  
    const rows: any[] = [];
    

    latestValue.results.forEach((item) => {
      const {name} = item;
      const {value, created_at: data} = item.latest;

      rows.push(createData(name, value, format(new Date(data), 'dd/MM/yyyy HH:MM', { locale: ptBR })))
    });

    return (
      <Box marginTop="12px" overflow="auto" maxHeight="200px">
        <TableContainer component={Paper}>
        <Table sx={{ minWidth: 650 }} aria-label="simple table">
            <TableHead>
            <TableRow>
                <TableCell>{title}</TableCell>
                <TableCell align="right">Último valor</TableCell>
                <TableCell align="right">Última atualização</TableCell>
            </TableRow>
            </TableHead>
            <TableBody>
            {rows.map((row) => (
                <TableRow
                key={row.name}
                sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                >
                <TableCell component="th" scope="row">{row.name}</TableCell>
                <TableCell align="right">{row.latestValue}</TableCell>
                <TableCell align="right">{row.latestData}</TableCell>
                </TableRow>
            ))}
            </TableBody>
        </Table>
        </TableContainer>
      </Box>
      );
}

export default LatestValueTable;