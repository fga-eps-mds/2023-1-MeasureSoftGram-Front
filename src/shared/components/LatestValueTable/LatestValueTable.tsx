import React from 'react';
import { format } from 'date-fns';
import { ptBR } from 'date-fns/locale';
import {
  Box,
  TableContainer,
  Table,
  TableCell,
  TableHead,
  TableRow,
  TableBody,
  Paper,
  Alert,
  Fade,
  Skeleton
} from '@mui/material';

import { useRequestLatestValues } from '@hooks/useRequestLatestValues';
import _ from 'lodash';

interface Prop {
  title: string;
  value: 'characteristics' | 'subcharacteristics' | 'measures' | 'metrics';
}

function LatestValueTable({ title, value }: Prop) {
  const { data, error, isLoading, isEmpty } = useRequestLatestValues(value);
  const tableRows: { name: string; latestValue: number; latestData: string }[] = [];

  _.forEach(data, (item) => {
    const { name } = item;
    const { value: latestValue, created_at: latestData } = item.latest;

    tableRows.push({
      name,
      latestValue,
      latestData: format(new Date(latestData), 'dd/MM/yyyy HH:mm:ss', { locale: ptBR })
    });
  });

  return isLoading ? (
    <Skeleton sx={{ width: '100%', minWidth: 400, height: '215px' }} />
  ) : (
    <Box marginTop="12px">
      <TableContainer component={Paper} sx={{ minWidth: 350, maxHeight: '215px' }}>
        <Table stickyHeader aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>{title}</TableCell>
              <TableCell align="right">Último valor</TableCell>
              <TableCell align="right">Última medição</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {tableRows.map((row) => (
              <TableRow key={row.name} sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
                <TableCell component="th" scope="row">
                  {row.name}
                </TableCell>
                <TableCell align="right">{row.latestValue.toFixed(2)}</TableCell>
                <TableCell align="right" sx={{ width: '250px' }}>
                  {row.latestData}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      {error && (
        <Fade in>
          <Alert variant="standard" severity="error">
            Ocorreu um erro ao tentar carregar as informações
          </Alert>
        </Fade>
      )}
      {isEmpty && (
        <Fade in>
          <Alert variant="standard" severity="warning">
            Não há dados para serem exibidos
          </Alert>
        </Fade>
      )}
    </Box>
  );
}

export default LatestValueTable;
