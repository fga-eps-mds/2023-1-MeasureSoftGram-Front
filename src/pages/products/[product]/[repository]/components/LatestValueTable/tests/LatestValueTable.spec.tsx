import React from 'react';
import { render } from '@testing-library/react';
import LatestValueTable from '../LatestValueTable';

describe('LatestValueTable', () => {
  it('renders with latestValue', () => {
    const title = 'TÃ­tulo da tabela';
    const latestValue = {
      results: [
        {
          name: 'item 1',
          latest: {
            value: 123,
            created_at: '2022-12-01T09:12:00.000Z'
          }
        },
        {
          name: 'item 2',
          latest: {
            value: 456,
            created_at: '2022-12-02T09:12:00.000Z'
          }
        }
      ]
    };
    const { container } = render(<LatestValueTable title={title} latestValue={latestValue} />);
    expect(container).toMatchSnapshot();
  });
});
