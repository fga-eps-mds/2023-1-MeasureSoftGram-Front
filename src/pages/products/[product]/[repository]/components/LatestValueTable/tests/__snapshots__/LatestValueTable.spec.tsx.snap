// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LatestValueTable renders with latestValue 1`] = `
<div>
  <div
    class="MuiBox-root css-14404vm"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="simple table"
        class="MuiTable-root css-quj9j6-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Título da tabela
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
              scope="col"
            >
              Último valor
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
              scope="col"
            >
              Última atualização
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-34nofg-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              scope="row"
            >
              item 1
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              123
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              01/12/2022 09:12
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-34nofg-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              scope="row"
            >
              item 2
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              456
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              02/12/2022 09:12
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
